src/adder.rs

```diff
@@ -2,35 +2,62 @@ mod table_header_adder;
 
 use anyhow::{bail, Context, Result};
 use serde_json::{from_str, Value as JValue};
-use toml_edit::{Array, ArrayOfTables, DocumentMut, InlineTable, Item, Table, Value};
+use toml_edit::{Array, ArrayOfTables, DocumentMut, InlineTable, Item, Value};
 
 use crate::converter::json_to_toml;
 use crate::field_finder::{get_field, DoInsert, TomlValue};
+use crate::AddOp;
 
-pub fn handle_add(
-    field: &str,
-    table_header_path: Option<String>,
-    dotted_path: Option<String>,
-    value: &str,
-    doc: &mut DocumentMut,
-) -> Result<()> {
-    match table_header_path {
+pub fn handle_add(doc: &mut DocumentMut, op: AddOp) -> Result<()> {
+    let path = op.dotted_path.or(op.path); // TODO: dotted_path is just a duplicated
+                                           // codepath of "path". Delete this once pid1 has
+                                           // been updated.
+    match op.table_header_path {
         Some(thpath) => {
-            let dpath = dotted_path.context("error: expected dotted_path to add")?;
-            handle_add_with_table_header_path(&thpath, &dpath, value, doc)
+            let value = op.value.context("error: expected value to add")?;
+            let mut table_header_path_vec = thpath
+                .split('/')
+                .map(|s| s.to_string())
+                .collect::<Vec<String>>();
+            let dotted_path_vec =
+                path.map(|p| p.split('/').map(|s| s.to_string()).collect::<Vec<String>>());
+            let field_value_json: JValue =
+                from_str(&value).context("parsing value field in add request")?;
+            let field_value_toml: Item = json_to_toml(&field_value_json, true)
+                .context("converting value in add request from json to toml")?;
+
+            let array_of_tables = if table_header_path_vec
+                .last()
+                .is_some_and(|key| key == "[[]]")
+            {
+                table_header_path_vec.pop();
+                true
+            } else {
+                false
+            };
+            table_header_adder::add_value_with_table_header_and_dotted_path(
+                doc,
+                &table_header_path_vec,
+                dotted_path_vec,
+                field_value_toml,
+                array_of_tables,
+            )
         }
         None => {
-            let mut path_split = field
+            let mut path_split = path
+                .expect("Missing 'path' value")
```

Comment:
Should we just return an error instead of panicking?

Reply:
Done!